Source/Common/Profiler.cpp

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: MIT
+#include "Common/Profiler.h"
+#include "git_version.h"
+
+#include <FEXCore/Debug/InternalThreadState.h>
+
+namespace FEX::Profiler {
+void StatAllocBase::SaveHeader(FEXCore::Profiler::AppType AppType) {
+  if (!Base) {
+    return;
+  }
+
+  Head = reinterpret_cast<FEXCore::Profiler::ThreadStatsHeader*>(Base);
+  Head->Size.store(CurrentSize, std::memory_order_relaxed);
+  Head->Version = FEXCore::Profiler::STATS_VERSION;
+
+  constexpr std::array<char, std::char_traits<char>::length(GIT_DESCRIBE_STRING) + 1> GitString = {GIT_DESCRIBE_STRING};
+  strncpy(Head->fex_version, GitString.data(), std::min(GitString.size(), sizeof(Head->fex_version)));
+  Head->app_type = AppType;
+
+  Stats = reinterpret_cast<FEXCore::Profiler::ThreadStats*>(reinterpret_cast<uint64_t>(Base) + sizeof(FEXCore::Profiler::ThreadStatsHeader));
+
+  RemainingSlots = TotalSlotsFromSize();
+}
+
+bool StatAllocBase::AllocateMoreSlots() {
+  const auto OriginalSlotCount = TotalSlotsFromSize();
+
+  uint64_t NewSize = AllocateMoreSlots(CurrentSize * 2);
+
+  if (NewSize == CurrentSize) {
+    return false;
+  }
+
+  CurrentSize = NewSize;
+  Head->Size.store(CurrentSize, std::memory_order_relaxed);
+  RemainingSlots = TotalSlotsFromSize() - OriginalSlotCount;
+
+  return true;
+}
+
+FEXCore::Profiler::ThreadStats* StatAllocBase::AllocateBaseSlot(uint32_t TID) {
+  if (!RemainingSlots) {
+    if (!AllocateMoreSlots()) {
+      return nullptr;
+    }
+  }
+
+  // Find a free slot
+  memory_barrier();
```

Comment:
Compared to the cost of the profiled events (SIGBUS/SMC), is it really worth dealing with the added complexity here instead of just using stronger atomics? Also, isn't thread creation/destruction in particular rare enough that we could use strong memory ordering on slot management variables while leaving the stats themselves weakly ordered?

Reply:
False dependency sharing in the hot path is a significant concern, so I'm keeping them as weak atomics.

Reply:
Where's the hot path? The currently traced events are orders of magnitude more expensive to process than a potential cache miss.

This applies even more to thread creation/destruction.

Reply:
Yes, we have some leniency on thread creation and teardown, which is already slow and causes global locks to be taken, but this is the way I have decided to implement it.